core/trino-main/src/main/java/io/trino/server/testing/exchange/LocalFileSystemExchange.java

```diff
@@ -0,0 +1,283 @@
+    {
+        synchronized (this) {
+            noMoreSinks = true;
+        }
+        checkInputReady();
+    }
+
+    @Override
+    public ExchangeSinkInstanceHandle instantiateSink(ExchangeSinkHandle sinkHandle, int taskAttemptId)
+    {
+        LocalFileSystemExchangeSinkHandle localFileSystemSinkHandle = (LocalFileSystemExchangeSinkHandle) sinkHandle;
+        Path outputDirectory = getExchangeDirectory()
+                .resolve(Integer.toString(localFileSystemSinkHandle.getTaskPartitionId()))
+                .resolve(Integer.toString(taskAttemptId));
+        return new LocalFileSystemExchangeSinkInstanceHandle(localFileSystemSinkHandle, outputDirectory, outputPartitionCount);
+    }
+
+    @Override
+    public void sinkFinished(ExchangeSinkInstanceHandle handle)
+    {
+        synchronized (this) {
+            LocalFileSystemExchangeSinkInstanceHandle localHandle = (LocalFileSystemExchangeSinkInstanceHandle) handle;
+            finishedSinks.add(localHandle.getSinkHandle());
+        }
+        checkInputReady();
+    }
+
+    private void checkInputReady()
+    {
+        verify(!Thread.holdsLock(this));
+        List<ExchangeSourceHandle> exchangeSourceHandles = null;
+        synchronized (this) {
+            if (exchangeSourceHandlesCreated) {
+                return;
+            }
+            if (noMoreSinks && finishedSinks.containsAll(allSinks)) {
+                // input is ready, create exchange source handles
+                exchangeSourceHandles = createExchangeSourceHandles();
+                exchangeSourceHandlesCreated = true;
+            }
+        }
+        if (exchangeSourceHandles != null) {
+            exchangeSourceHandlesFuture.complete(exchangeSourceHandles);
+        }
+    }
+
+    private synchronized List<ExchangeSourceHandle> createExchangeSourceHandles()
+    {
+        Multimap<Integer, String> partitionFilesMap = ArrayListMultimap.create();
+        for (LocalFileSystemExchangeSinkHandle sinkHandle : finishedSinks) {
+            Path committedAttemptPath = getCommittedAttemptPath(sinkHandle);
+            Map<Integer, Path> partitions = getCommittedPartitions(committedAttemptPath);
+            partitions.forEach((partition, file) -> partitionFilesMap.put(partition, file.toAbsolutePath().toString()));
+        }
+
+        ImmutableList.Builder<ExchangeSourceHandle> result = ImmutableList.builder();
+        for (Integer partitionId : partitionFilesMap.keySet()) {
+            result.add(new LocalFileSystemExchangeSourceHandle(partitionId, ImmutableList.copyOf(partitionFilesMap.get(partitionId))));
+        }
+        return result.build();
+    }
+
+    private Path getCommittedAttemptPath(LocalFileSystemExchangeSinkHandle sinkHandle)
+    {
+        Path sinkOutputBasePath = getExchangeDirectory()
+                .resolve(Integer.toString(sinkHandle.getTaskPartitionId()));
+        try {
+            List<Path> attemptPaths = Files.list(sinkOutputBasePath)
```

Comment:
After a lot of debugging of "too-many-open-files" errors on Mac I learned that `Files.list` cannot be used like that.
There is a note in javadoc:
```
API Note:
This method must be used within a try-with-resources statement or similar control structure to ensure that the stream's open directory is closed promptly after the stream's operations have completed.
```

Workaround which fixes issue is to add:
```java
    private static List<Path> listFiles(Path directory, Predicate<Path> predicate)
            throws IOException
    {
        ImmutableList.Builder<Path> builder = ImmutableList.builder();
        try (Stream<Path> dir = Files.list(directory)) {
            dir.filter(predicate).forEach(builder::add);
        }
        return builder.build();
    }
```

And then replace call spots with:
```
List<Path> attemptPaths = listFiles(sinkOutputBasePath, Files::isDirectory);
```

and 

```
List<Path> partitionFiles = listFiles(committedAttemptPath, path -> path.toString().endsWith(DATA_FILE_SUFFIX));
```

cc: @linzebing. I did not check if the issue applies to https://github.com/trinodb/trino/pull/10376 - but there is a chance it does.

Reply:
Another good catch, thanks @losipiuk